### ISSUE TITLE
Conversion from unsafe.Pointer to *V violates unsafe package rules 

### ISSUE THREAD

**Issue by DmitriyMV** [smallnest/exp#3]
https://github.com/smallnest/exp/blob/3592a45ffb605184a355afc1807b4a6d41cc1f4d/sync/sync_map.go#L194
https://github.com/smallnest/exp/blob/3592a45ffb605184a355afc1807b4a6d41cc1f4d/sync/sync_map.go#L519

I got curious about this approach and talked with Go devs and community - is it safe to do so? Their answer was **no** - it's not. Go compiler makes no promises about not following pointer value and exploring contents in the runtime.

Per [docs](https://pkg.go.dev/unsafe#Pointer):
```
(1) Conversion of a *T1 to Pointer to *T2.

Provided that T2 is no larger than T1 and that the two share an equivalent memory layout, this conversion allows reinterpreting data of one type as data of another type. An example is the implementation of math.Float64bits:
```

So you can't do this technically. What you can do is use plain unsafe.Pointer instead of atomic.Pointer.

**Comment by smallnest** [comment]
Yes, theoretically it seems `(*V)(expunged)` violates unsafe package rules as you mentioned，but here it is trick that we use the expunged as immutable and we don't change its bytes. It only acts as a special value to indicate values in map are deleted, so I personally think it is not a problem.

Why not use unsafe.Pointer? I hope this code  can keep up to changes in go official repository.
There is a repository to use unsafe.Pointer https://github.com/go-typ/typ/blob/main/sync2/map.go#L78
